model-optimizer/extensions/ops/Cast.py

```diff
@@ -47,6 +49,60 @@ def type_infer(node: Node):
         assert node.has_valid('dst_type'), 'Destination type of "Cast" operation should be extracted earlier'
         node.out_port(0).set_data_type(node.dst_type)
 
+    @staticmethod
+    def helper_value_propagation(node_name, value, dst_type):
+        new_blob, finite_match_count, zero_match_count = convert_blob(value, dst_type)
+
+        if finite_match_count:
+            log.error("{} elements of {} were clipped to infinity while converting an input blob for node '{}' to {}."
+                      " ".format(finite_match_count, new_blob.size, node_name, dst_type) + refer_to_faq_msg(76))
+        if zero_match_count:
+            log.warning("{} elements of {} were clipped to zero while converting an input blob for node '{}' to {}."
+                        " ".format(zero_match_count, new_blob.size, node_name, dst_type) + refer_to_faq_msg(77))
+        return new_blob
+
+    @staticmethod
+    def custom_type_casting_and_packing(node: Node, value, dst_type):
+        """
+        Custom types are not supported by numpy but we still need to write it to the .bin file in a compact way.
+        To do so we prepare bit representation of int4/uint4 values and store them in a numpy friendly data type.
+        We pack int4/uint4 values into uint8 type (two int4/uint4 numbers fit in uint8).
+        If the number of elements in the blob is uneven we pad them with zero value to be able to fit the bit sequence
+        into the uint8 array.
+        Example: we need to represent 5 elements of int4 dtype
+            we would pad them to 6 element with the last element as zero and we would pack them into 3 uint8 values
+        """
+        assert dst_type in [packed_U4, packed_I4]
+
+        minimum_regular_dtype = np.uint8 if dst_type == packed_U4 else np.int8
+        casted_to_regular_type = Cast.helper_value_propagation(
+            node.soft_get('name', node.id), value, minimum_regular_dtype)
+
+        data_shape = node.out_port(0).data.get_shape().copy()
```

Comment:
If the result of `node.out_port(0).data.get_shape()` is `None` then we will crash here